### PR TITLE
Optimize SP1 Prover Build Time by Caching Proving and Verifying Keys

### PR DESCRIPTION
**Pull request by prajwolrg**
## Description

This PR introduces a caching mechanism for the proving and verifying keys in the SP1 prover build process. By caching these keys and checking for changes to the ELF file, we can avoid recalculating the keys on every build, reducing build time by approximately half on subsequent builds.

### build.rs Update:

Added functions to check for cache validity based on the SHA-256 hash of the ELF file.
Modified the build logic to reuse cached Proving Key (PK) and Verifying Key (VK) when no changes are detected.
Improved error handling and logging to provide clearer information on cache operations.

### New Cache Directory:

Created a cache directory for each guest program in the guests folder.
Four new files are stored per program:
- `guest_program_name.elf`: Compiled ELF file.
- `guest_program_name.id`: SHA-256 hash of the ELF file.
- `guest_program_name.pk`: Cached Proving Key.
- `guest_program_name.vk`: Cached Verifying Key.

### Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [X] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [X] Refactor
-   [ ] New or updated tests

## Checklist

-   [ ] I have performed a self-review of my code.
-   [X] I have commented my code where necessary.
-   [X] I have updated the documentation if needed.
-   [X] My changes do not introduce new warnings.
-   [ ] I have added tests that prove my changes are effective or that my feature works.
-   [ ] New and existing tests pass with my changes.

## Related Issues

Built on top of #448

### CODE REVIEW

File: provers/sp1/build.rs [hunk 1]
```diff
@@ -61,18 +83,27 @@ pub const {}: &str = "{}";
...
+    // Write the accumulated methods_file_content to methods.rs in the output directory
     let out_dir = get_output_dir();
     let methods_path = out_dir.join("methods.rs");
-    fs::write(&methods_path, methods_file_content).expect("Failed writing to methods path");
+    fs::write(&methods_path, methods_file_content).unwrap_or_else(|e| {
+        panic!(
+            "Failed to write methods.rs file at {}: {}",
+            methods_path.display(),
+            e
+        )
```
Comment (storopoli) [inline provers/sp1/build.rs:94]: Do we want to handle gracefully this error or just crash and burn?

File: provers/sp1/build.rs [hunk 2]
```diff
@@ -96,19 +127,21 @@ fn build_program_with_dependencies(
...
 
-        // Only write vks.rs if there are dependencies and is being built in release mode
-        #[cfg(not(debug_assertions))]
+        // Only write vks.rs if there are dependencies
         if !vks_content.is_empty() {
-            // Write vks.rs in the program's directory before building it
             let vks_path = Path::new(program).join("src").join("vks.rs");
-            fs::write(&vks_path, vks_content).expect("Failed writing to vks.rs");
+            fs::write(&vks_path, vks_content)
+                .unwrap_or_else(|e| panic!("Failed to write vks.rs for {}: {}", program, e));
```
Comment (storopoli) [inline provers/sp1/build.rs:134]: Same here, crash and burn?

File: provers/sp1/build.rs [hunk 3]
```diff
@@ -117,30 +150,87 @@ fn build_program_with_dependencies(
...
-//     env::var("SKIP_GUEST_BUILD").is_err() && env::var("CARGO_CFG_CLIPPY").is_err()
-// }
-
+/// Returns the output directory for the build artifacts.
 fn get_output_dir() -> PathBuf {
     env::var_os("OUT_DIR")
-        .expect("OUT_DIR environment variable not set")
-        .into()
+        .map(PathBuf::from)
+        .expect("OUT_DIR environment variable is not set. Cannot determine output directory.")
```
Comment (storopoli) [inline provers/sp1/build.rs:157]: Can't we gracefully hard-code this to a fallback `OUT_DIR`?

File: provers/sp1/build.rs [hunk 4]
```diff
@@ -117,30 +150,87 @@ fn build_program_with_dependencies(
...
+            deserialize(&serialized_vk).map_err(|e| format!("VK deserialization failed: {}", e))?;
+        Ok((elf, vk))
+    }
+}
+
+/// Generates the ELF contents and VK hash for a given program.
 #[cfg(not(debug_assertions))]
 fn generate_elf_contents_and_vk_hash(program: &str) -> (Vec<u8>, [u32; 8], String) {
     // Setup compiler
+    const RISC_V_COMPILER: &str = "/opt/riscv/bin/riscv-none-elf-gcc";
```
Comment (storopoli) [inline provers/sp1/build.rs:218]: Is this path also the same in Linux. This feels MacOS-y...

File: provers/sp1/build.rs [hunk 5]
```diff
@@ -117,30 +150,87 @@ fn build_program_with_dependencies(
...
+    // Build the program
+    build_program_with_args(program, build_args);
 
-    let contents = fs::read(elf_path).expect("Failed to find SP1 ELF");
-    let client = MockProver::new();
-    let (_pk, vk) = client.setup(&contents);
-    (contents, vk.hash_u32(), vk.bytes32())
+    // Now, ensure cache validity
+    let (elf, vk) = ensure_cache_validity(program)
+        .expect("Failed to ensure cache validity after building program");
```
Comment (storopoli) [inline provers/sp1/build.rs:232]: Same here it will crash and burn.

**Review by storopoli — COMMENTED** [review]
Some comments on the `expect` and `panic`s

File: provers/sp1/build.rs (hunk 1 above)
Comment (prajwolrg) [reply provers/sp1/build.rs:94]: Since this is a build file, imo crashing the build makes more sense.

File: provers/sp1/build.rs (hunk 2 above)
Comment (prajwolrg) [reply provers/sp1/build.rs:134]: same as above.

File: provers/sp1/build.rs (hunk 3 above)
Comment (prajwolrg) [reply provers/sp1/build.rs:157]: `OUT_DIR` is used in [`src/main.rs`](https://github.com/alpenlabs/strata/blob/main/provers/sp1/src/lib.rs). Having a fallback `OUT_DIR` might increase the complexity.

File: provers/sp1/build.rs (hunk 4 above)
Comment (prajwolrg) [reply provers/sp1/build.rs:218]: Oh yes right! It's not required with the latest version. Created PR #448 to move to latest version.

File: provers/sp1/build.rs (hunk 5 above)
Comment (prajwolrg) [reply provers/sp1/build.rs:232]: yes.

File: provers/sp1/build.rs (hunk 1 above)
Comment (storopoli) [reply provers/sp1/build.rs:94]: Ok, makes sense. Thanks!

**Review by MdTeach — APPROVED** [review]

**Pull request merged**